### PR TITLE
CRM-14536 fix for master

### PR DESCRIPTION
**Pull request by monishdeb**
---
- CRM-14536: Free Membership still submits to payment processor
  https://issues.civicrm.org/jira/browse/CRM-14536

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
This looks like an improvement - since it removes code duplication!!! Yay.

My concern is that the BilllingBlock.tpl should be loaded directly or indirectly by ALL forms that include payment fields. The changes here apply to front end but not back end forms, a distinction we've been trying to remove at the payment level.

The billing fields are excluded by virtue of the tpl checking if not required - so the tpl should be included regardless of the processor or the form. ie.

 {if $paymentFields|@count}
{if $billingDetailsFields|@count}

The goal with the paypal js is that it would be included from the CRM_Core_Payment_PayPalImpl::buildForm method. Already some js is being added from here & some tpl code - I'm not sure whether we can just move these paypal bits of js into there as well or if we are still hitting issues around the timing of the inclusion of the js.

The end goal is that any payment processor specific code is driven by the class in question & the paypal js is the main area where that is still not happening.

Regarding pay later - there is no reason not to include billing.tpl for paylater - which should also only show the fields if they are assigned & if billing fields are required for tax purposes php can drive that by assigning or not :
$billingDetailsFields

**Comment by eileenmcnaughton** [comment]
@colemanw you might like to comment

**Comment by monishdeb** [comment]
As per the current Js changes that will be applied to - 
1. toggle payment option on basis of $0 membership/contribution payment on choosing quick-config/price-set fields.
NOTE: It can be moved to templates/CRM/common/paymentBlock.tpl along with including payment option form elements 
 
2. paypal specific code is only related to Paypal express pp as we show express button instead of "Confirm Contribution" submit button
 NOTE : It is possible to move paypal specific js (although existing code is not optimized enough as showHidePayPalExpressOption() and  toggleConfirmButton() shares the same logic except pay-later factor) to paypal script.

3. Moving toggleConfirmButton() fn from BillingBlock.tpl to the new tpl file as of now BillingBlock.tpl snippet not get loaded for some cases like in paypal-express or choosing pay-later there is no need for it. (In a dilemma whether we should always include billing block for each pp by making it non-/required respectively)

**Comment by eileenmcnaughton** [comment]
Re 3 - I'm actually strongly of the opinion billingblock.tpl should always be included & if it's not always happening it's worth digging out the occasions when it isn't. The last thing we want is form code deciding not to included it because the processor is paypal express - since that goes against the logic we are working towards. Also, sometimes billing fields are required in pay later (for tax invoicing).

However, there should be no harm done if it is over-included, as it won't display the fields if they are not assigned to the template

File: CRM/Contribute/BAO/Contribution/Utils.php [hunk 1]
```diff
@@ -118,13 +118,17 @@ public static function processConfirm(
...
+      $pending = (empty($paymentParams['is_pay_later']) && $form->_amount > 0.0) ? TRUE : FALSE;
+
       $contribution = CRM_Contribute_Form_Contribution_Confirm::processFormContribution(
         $form,
         $paymentParams,
         NULL,
         $contributionParams,
         $financialType,
-        TRUE,
+        $pending,
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution/Utils.php]: I am 95% of the way to removing pending as a function paramter for processFormContribution - with the idea that they are always pending & confirmed via completetransaction if paid

**Pull request closed without merging**